src/vmspawn/vmspawn-scope.c

```diff
@@ -0,0 +1,306 @@
+
+        FOREACH_ARRAY(cmd, commands, n_commands) {
+                char **cmdline = *cmd;
+
+                r = sd_bus_message_open_container(m, 'r', "sasb");
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_append(m, "s", cmdline[0]);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_append_strv(m, cmdline);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_append(m, "b", 0);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_close_container(m);
+                if (r < 0)
+                        return bus_log_create_error(r);
+        }
+
+        r = sd_bus_message_close_container(m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_close_container(m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        r = sd_bus_message_close_container(m);
+        if (r < 0)
+                return bus_log_create_error(r);
+
+        return 0;
+}
+
+/* NOTE: this function should really take cmdline as `const char * const *` but cannot until sd_bus_message_append_str does as well */
+int attach_command_to_socket_in_scope(sd_bus *bus, const char *scope, const char *unit_name_prefix, const char *socket_path,
+                                      int socket_type, char **cmdline, char **cleanup_cmdline, char **extra_properties) {
+        _cleanup_(bus_wait_for_jobs_freep) BusWaitForJobs *w = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *m = NULL, *reply = NULL;
+        _cleanup_free_ char *service_desc = NULL, *service_name = NULL, *socket_name = NULL, *socket_desc = NULL;
+        const char *object, *socket_type_str;
+        int r;
+
+        /*
+         * Starts a socket/service unit pair bound to the given scope.
+         *
+         * The socket unit listens on `socket_path` for connections of type `socket_type`.
+         * When a connection to the socket is initiated the service is started as well and passed the socket's file descriptor as fd 3.
+         *
+         * `cmdline` is forwarded to the service unit's ExecStart property.
+         * `cleanup_cmdline` is forwarded to the service unit's ExecStopPost property.
+         *
+         * `extra_properties` is used to set any extra properties this service unit should have.
+         * The format is the same as in unit files, e.g. Requires=other-unit.socket
+         */
+
+        assert(bus);
+        assert(scope);
+        assert(unit_name_prefix);
+        assert(socket_path);
+        assert(cmdline);
+        assert_return(IN_SET(socket_type, SOCK_STREAM, SOCK_SEQPACKET, SOCK_DGRAM), -EOPNOTSUPP);
```

Comment:
We don't really use `assert_return()` in internal functions. Please make this a regular if and log an error if it fails.

Reply:
I'm not sure we actually need the assertion at all, theres nothing wrong with the other socket types and I think this was an artifact of how I was turning it into a string. As I now use a table for this it I think we can drop the assert?